### ISSUE TITLE
Unable to upgrade PostgreSQL from 9.5 to 9.6 with pg_pathman 1.2.1 installed

### ISSUE THREAD

**Issue by Envek** [postgrespro/pg_pathman#75]
I'm trying to upgrade PostgreSQL with pg_pathman-enabled database using `pg_upgrade`. Both new and old instances have most recent pg_pathman 1.2.1 installed and added to `postgresql.conf`.

Invoking `pg_upgrade`:

```
postgres@smp-bsktest-db-0:/var/tmp/pg_upgrade$ /usr/lib/postgresql/9.6/bin/pg_upgrade --check --link --jobs `nproc` --old-bindir /usr/lib/postgresql/9.5/bin --new-bindir /usr/lib/postgresql/9.6/bin --old-datadir /var/lib/postgresql/9.5/main --new-datadir /var/lib/postgresql/9.6/main --old-options ' -c config_file=/etc/postgresql/9.5/main/postgresql.conf -c hba_file=/etc/postgresql/9.5/main/pg_hba.conf' --new-options ' -c config_file=/etc/postgresql/9.6/main/postgresql.conf -c hba_file=/etc/postgresql/9.6/main/pg_hba.conf'

*failure*
Consult the last few lines of "pg_upgrade_server.log" for
the probable cause of the failure.
Performing Consistency Checks on Old Live Server
------------------------------------------------
Checking cluster versions                                   ok
Checking database user is the install user                  ok
Checking database connection settings                       ok
Checking for prepared transactions                          ok
Checking for reg* system OID user data types                fatal

Your installation contains one of the reg* data types in user tables.
These data types reference system OIDs that are not preserved by
pg_upgrade, so this cluster cannot currently be upgraded.  You can
remove the problem tables and restart the upgrade.  A list of the problem
columns is in the file:
    tables_using_reg.txt

Failure, exiting
```

Contents of `tables_using_reg.txt`:
```
Database: smp
  public.pathman_config_params.init_callback
```

Is it (or will be) possible to upgrade PostgreSQL? How?

Thank you in advance.

**Comment by funbringer** [comment]
Hi @Envek,

The inability to upgrade is caused by the `regprocedure` column of `pathman_config_params`. Currently it's impossible to perform an upgrade, since PostgreSQL doesn't preserve OIDs of functions. We're sorry for the inconvenience :(

We'll definitely find a solution. For instance, we could change the column's type.

**Comment by maksm90** [comment]
Hi, @Envek !

I have fixed the problem with migration of pathman's configure tables. In particular, I have changed type of problem column `init_callback` from `regprocedure` to `text`.

All changes have been committed in the branch `master_pg_upgrade_bug`. Looking forward to your possible testing and giving any feedback.

**Comment by Envek** [comment]
@maksm90, thank you and sorry for the delay.

I successfully upgraded cluster with fresh installation of pg_pathman from `master_pg_upgrade_bug` with fresh db.

But the problem is current installation: just upgrading `pg_pathman` is not enough as column `init_callback` column should be somehow migrated from `regprocedure` to `text` in existing installations. In my test DB I was need to drop partitioning and setup it again.

Is there any way to migrate this column?

If this fix will go into 1.3 release, please add migration of this column's type in update scripts, so I will be able to upgrade PostgreSQL after installing new pg_pathman and executing `ALTER EXTENSION pg_pathman UPDATE TO '1.3';`. 

When I was tried to reapply migration, I've got following error at the end:

    SELECT partition_table_concurrently('tracker_points')")
    PG::InternalError: ERROR:  cannot start ConcurrentPartWorker
    DETAIL:  table is being partitioned now

and transaction rollbacks.

The migration is as follows (it was originally written for pg_pathman 1.1):

```sql
CREATE EXTENSION pg_pathman;
SELECT create_range_partitions( 'tracker_points', 'evented_at', '2015-06-01'::timestamp, '1 month'::interval, 0, false );
SELECT add_range_partition( 'tracker_points', '2015-06-01'::timestamp, '2015-06-01'::timestamp + '1 month'::interval, 'tracker_points_2015_06' );
SELECT append_range_partition('tracker_points', 'tracker_points_2015_07');
-- and so on
SELECT append_range_partition('tracker_points', 'tracker_points_2017_01');
SELECT set_enable_parent('tracker_points', false);
SELECT partition_table_concurrently('tracker_points');
```

**Comment by maksm90** [comment]
@Envek, to disable/enable `pg_pathman` you may not delete all partitions and after that restore their. Suffice to cascade drop `pg_pathman` and execute function `add_to_pathman_config` on partitioned tables after restoring new version of `pg_pathman`. Obviously, you have to restore your specific partition settings by hand.

Release `pg_pathman` 1.3 will be soon (on the next week) and the fix of callback's column type will be inside it. We'll certainly prepare migration script including the change of callback's column type.

As for the error with `partition_table_concurrently`, I'll try to reproduce your situation and get outside of a problem in the very near future.

**Comment by funbringer** [comment]
Fixed in [release 1.3](https://github.com/postgrespro/pg_pathman/releases/tag/1.3):

```
Performing Consistency Checks
-----------------------------
Checking cluster versions                                   ok
Checking database user is the install user                  ok
Checking database connection settings                       ok
Checking for prepared transactions                          ok
Checking for reg* system OID user data types                ok
Checking for contrib/isn with bigint-passing mismatch       ok
Checking for roles starting with 'pg_'                      ok
Creating dump of global objects                             ok
Creating dump of database schemas
                                                            ok
Checking for presence of required libraries                 ok
Checking database user is the install user                  ok
Checking for prepared transactions                          ok

If pg_upgrade fails after this point, you must re-initdb the
new cluster before continuing.

Performing Upgrade
------------------
Analyzing all rows in the new cluster                       ok
Freezing all rows on the new cluster                        ok
Deleting files from new pg_clog                             ok
Copying old pg_clog to new server                           ok
Setting next transaction ID and epoch for new cluster       ok
Deleting files from new pg_multixact/offsets                ok
Copying old pg_multixact/offsets to new server              ok
Deleting files from new pg_multixact/members                ok
Copying old pg_multixact/members to new server              ok
Setting next multixact ID and offset for new cluster        ok
Resetting WAL archives                                      ok
Setting frozenxid and minmxid counters in new cluster       ok
Restoring global objects in the new cluster                 ok
Restoring database schemas in the new cluster
                                                            ok
Copying user relation files
                                                            ok
Setting next OID for new cluster                            ok
Sync data directory to disk                                 ok
Creating script to analyze new cluster                      ok
Creating script to delete old cluster                       ok

Upgrade Complete
----------------
Optimizer statistics are not transferred by pg_upgrade so,
once you start the new server, consider running:
    ./analyze_new_cluster.sh

Running this script will delete the old cluster's data files:
    ./delete_old_cluster.sh

```

**Comment by Envek** [comment]
Confirming: I have successfully upgraded too. Thank you!
